fdbserver/TLogServer.actor.cpp

```diff
@@ -2107,7 +2115,7 @@ Future<Void> tLogPeekMessages(PromiseType replyPromise,
 	auto messagesValue = messages.toValue();
 	reply.arena.dependsOn(messagesValue.arena());
 	reply.messages = messagesValue;
-	reply.end = endVersion;
+	reply.end = clusterRecoveryVersion.present() ? clusterRecoveryVersion.get() : endVersion;
```

Comment:
We need additional check here, because it's possible that after `endVersion`, there are data not sent yet. For instance, `peekMessagesFromMemory()` can return early if the message size exceeds `SERVER_KNOBS->DESIRED_TOTAL_BYTES`.

Reply:
> I do not think that is possible, because clusterRecoveryVersion is only set if the peek's begin version is greater than logData->version.get()
Cluster recovery version is the global recovery version (max of the recovery versions of all logs) if unicast is enabled, right? If so, shouldn't it be set always during recovery (instead of only in the case that you mentioned above)?

Reply:
>And another comment regarding cluster recovery version: it appears we never reset it on a locked log, is that going to be a problem if another recovery starts before the current recovery completes?

discussed / clarified code offline.

Reply:
Note that the clusterRecoveryVersion is set only if the request's beginVersion is greater than the last version received by the tLog.